### PR TITLE
feat: identify vulns introduced by the dockerfile

### PR DESCRIPTION
**Pull request by robcresswell**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -14,22 +14,25 @@
...
     "prepare": "npm run build"
   },
   "dependencies": {
+    "@types/node": "^8",
     "debug": "^3",
     "dockerfile-ast": "0.0.12",
     "semver": "^5.6.0",
-    "tslib": "^1"
+    "tslib": "^1",
+    "typescript": "^3"
```
Comment (darscan) [inline package.json]: Why have `typescript` and `@types/node` become prod deps?

File: lib/index.ts [hunk 2]
```diff
@@ -23,16 +26,55 @@ function inspect(root: string, targetFile?: string, options?: any) {
...
+
       return {
         plugin: metadata,
         package: pkg,
       };
     });
 }
 
+function getDockerfileDependencies(dockerfilePackages: DockerFilePackages,
+                                   dependencies) {
```
Comment (darmalovan) [inline lib/index.ts]: Can you add a type annotation to the return value of the function?

File: lib/index.ts [hunk 3]
```diff
@@ -23,16 +26,55 @@ function inspect(root: string, targetFile?: string, options?: any) {
...
+      };
+
       return {
         plugin: metadata,
         package: pkg,
       };
     });
 }
 
+function getDockerfileDependencies(dockerfilePackages: DockerFilePackages,
```
Comment (darmalovan) [inline lib/index.ts]: This function is more of `adding` the docker file package dependencies if I'm understanding it correctly.
Maybe `addDockerfilePackagesDependencies` ?

Maybe a small comment here explaining why we do this would be nice :)

File: test/lib/instruction-parser.spec.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
+import { getPackagesFromRunInstructions } from '../../lib/instruction-parser';
+
+describe('instruction parsers', () => {
+  describe('getPackagesFromRunInstructions', () => {
```
Comment (darmalovan) [inline test/lib/instruction-parser.spec.ts]: Can we add a test for empty instructions?

File: test/lib/instruction-parser.spec.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+      });
+    });
+
+    test('handles multiple spaces', () => {
+      const instruction = 'RUN    apt   install   curl   vim   ';
+      const packages = getPackagesFromRunInstructions([instruction]);
+
+      expect(packages).toEqual({
+        'curl': { instruction },
+        'vim': { instruction },
```
Comment (darmalovan) [inline test/lib/instruction-parser.spec.ts]: VIM 💪

File: test/lib/instruction-parser.spec.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+      const instruction = 'RUN apt install curl && apt install wget';
+      const packages = getPackagesFromRunInstructions([instruction]);
+
+      expect(packages).toEqual({
+        'curl': { instruction },
+        'wget': { instruction },
+      });
+    });
+
+    test('supports multiple commands using ";"', () => {
```
Comment (darmalovan) [inline test/lib/instruction-parser.spec.ts]: Nice ! can we add another test for this instruction:
`const instruction = 'RUN apt install curl; echo "rob"';`

**Comment by darmalovan** [comment]
Looks good! some not big comments

**Comment by darmalovan** [comment]
One thing I'm not sure about is the fact that a vulnerability can appear in multiple packages.
The difference will be the vulnerable path.
Our docker instruction output is per vulnerability and not per vulnerable path how it will work out?

For example the same vuln in `curl` and `wget`, let's `curl` is installed by the user - what will the CLI output will look like?

File: lib/instruction-parser.ts [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ * We also need to account for the multiple ways to split commands, and
+ * arbitrary whitespace
+ */
+function getPackagesFromRunInstructions(
+  runInstructions: string[],
+): DockerFilePackages {
+  return runInstructions.reduce((dockerfilePackages, instruction) => {
+    // A single RUN may include multiple commands, so remove "RUN " and split
+    const commands = instruction.slice(4).split(/\s?(;|&&)\s?/);
```
Comment (orkamara) [inline lib/instruction-parser.ts]: what's 4? the len of 'RUN '?
Maybe create a const and use length?

File: lib/instruction-parser.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
+export { getPackagesFromRunInstructions, DockerFilePackages };
+
+interface DockerFilePackages {
+  [packageName: string]: {
+    instruction: string;
+  };
+}
+
+// tslint:disable-next-line:max-line-length
+const installRegex = /\s*(rpm\s+-i|rpm\s+--install|apk\s+add|apt\s+install|apt-get\s+install)\s+(-\w\s+)?/;
```
Comment (orkamara) [inline lib/instruction-parser.ts]: Maybe we should also support more common ones? like `yum install` and `aptitude install`?

File: test/lib/instruction-parser.spec.ts [hunk 9]
```diff
@@ -0,0 +1,106 @@
+import { getPackagesFromRunInstructions } from '../../lib/instruction-parser';
+
+describe('instruction parsers', () => {
```
Comment (orkamara) [inline test/lib/instruction-parser.spec.ts]: Can we add another test for an install command spitted by multiple line?
`RUN apt-get update && apt-get install -y \
    python-qt4 \
    python3-pyqt5`

File: test/lib/instruction-parser.spec.ts [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    });
+
+    test('supports "rpm --install"', () => {
+      const instruction = 'RUN rpm --install curl';
+      const packages = getPackagesFromRunInstructions([instruction]);
+
+      expect(packages).toEqual({ 'curl': { instruction } });
+    });
+
+    test('ignores irrelevant flags', () => {
```
Comment (orkamara) [inline test/lib/instruction-parser.spec.ts]: can we add a test where the flags are in the end?
`apt-get install vim -y`

File: lib/index.ts [hunk 11]
```diff
@@ -23,16 +26,55 @@ function inspect(root: string, targetFile?: string, options?: any) {
...
+                                   dependencies) {
+  for (const dependencyName in dependencies) {
+    if (dependencies.hasOwnProperty(dependencyName)) {
+      const sourceOrName = dependencyName.split('/')[0];
+      const dockerfilePackage = dockerfilePackages[sourceOrName];
+
+      if (dockerfilePackage) {
+        collectDeps(dependencies[dependencyName]).forEach((dep) => {
```
Comment (darmalovan) [inline lib/index.ts]: Let's say package `a` is installed by the base image, and package `b` is installed by the docker file.
As I understand the code, it means that if a package `b` is depending on `a` it will mark `a` as a docker file package - which will mark a vuln on this package as a docker file vuln. LMK what do you think.

File: lib/instruction-parser.ts (hunk 7 above)
Comment (robcresswell) [reply lib/instruction-parser.ts]: Good point, will do

File: test/lib/instruction-parser.spec.ts (hunk 9 above)
Comment (robcresswell) [reply test/lib/instruction-parser.spec.ts]: Ah good suggestion! I don't think my code catches this.

File: lib/index.ts (hunk 11 above)
Comment (robcresswell) [reply lib/index.ts]: Yup, it will do exactly this. Any suggestions on improving it?

File: lib/index.ts (hunk 3 above)
Comment (robcresswell) [reply lib/index.ts]: Done

File: test/lib/instruction-parser.spec.ts (hunk 5 above)
Comment (robcresswell) [reply test/lib/instruction-parser.spec.ts]: YAAAASSSSS

File: test/lib/instruction-parser.spec.ts (hunk 6 above)
Comment (robcresswell) [reply test/lib/instruction-parser.spec.ts]: Updated the existing case to 

```
const instruction = 'RUN apt install curl; apt install vim; echo "bitwise"';
```

Hope that covers the case you were concerned about

File: lib/instruction-parser.ts (hunk 8 above)
Comment (robcresswell) [reply lib/instruction-parser.ts]: Done

File: test/lib/instruction-parser.spec.ts (hunk 10 above)
Comment (robcresswell) [reply test/lib/instruction-parser.spec.ts]: Good catch

File: lib/index.ts [hunk 12]
```diff
@@ -23,16 +26,62 @@ function inspect(root: string, targetFile?: string, options?: any) {
...
+
       return {
         plugin: metadata,
         package: pkg,
       };
     });
 }
 
+// Iterate over the dependencies list; if one is introduced by the dockerfile,
+// flatten it's dependencies and append them to the list of dockerfile
```
Comment (karniwl) [inline lib/index.ts]: sorry for nitpicking but `it's` should be `its` :)

Comment (robcresswell) [reply lib/index.ts]: Fixed

**Comment by darmalovan** [comment]
Approved, Just take a look at the test failures?
